File: wxFEFactory/pyapi/layout/containers.cpp
#include <wx/wx.h>
#include "containers.h"


/**
 * 获取布局参数
 */

void SizerLayout::getBoxArg(View & child, int * pFlex, int * pFlag, int * pPadding)
{
	int flag = 0;
	if (child.getStyle(STYLE_EXPAND, false))
	{
		flag |= wxEXPAND;
	}

	wxcstr showPad = child.getStyle(STYLE_SHOWPADDING, wxNoneString);

	if (showPad != wxNoneString)
	{
		if (showPad.size() == 1)
		{
			if (showPad[0] != '0')
				flag |= wxALL;
		}
		else if (showPad.size() == 7)
		{
			if (showPad[0] != '0')
				flag |= wxTOP;
			if (showPad[2] != '0')
				flag |= wxRIGHT;
			if (showPad[4] != '0')
				flag |= wxBOTTOM;
			if (showPad[6] != '0')
				flag |= wxLEFT;
		}
		else {
			log_message(wxString::Format(wxT("%s: %s not available"), STYLE_SHOWPADDING, showPad));
		}
	}

	wxcstr vertical = child.getStyle(STYLE_VERTICALALIGN, wxNoneString);

	if (vertical != wxNoneString)
	{
		if (vertical == wxT("top"))
		{
			flag |= wxALIGN_TOP;
		}
		else if (vertical == wxT("bottom"))
		{
			flag |= wxALIGN_BOTTOM;
		}
		else if (vertical == wxT("middle"))
		{
			flag |= wxALIGN_CENTER_VERTICAL;
		}
		else {
			log_message(wxString::Format(wxT("%s: %s not available"), STYLE_VERTICALALIGN, vertical));
		}
	}

	wxcstr align = child.getStyle(STYLE_ALIGN, wxNoneString);

	if (align != wxNoneString)
	{
		if (vertical == wxT("left"))
		{
			flag |= wxALIGN_LEFT;
		}
		else if (vertical == wxT("right"))
		{
			flag |= wxALIGN_RIGHT;
		}
		else if (vertical == wxT("center"))
		{
			flag |= wxALIGN_CENTER_HORIZONTAL;
		}
		else {
			log_message(wxString::Format(wxT("%s: %s not available"), STYLE_ALIGN, vertical));
		}
	}
	*pFlag = flag;
	*pFlex = child.getStyle(STYLE_FLEX, 0);
	*pPadding = child.getStyle(STYLE_PADDING, 5);
}

void SplitterWindow::__exit__(py::args & args)
{
	int len = m_children.size();
	if (len > 2)
	{
		log_message("SplitterWindow 不支持大于2个子元素");
		return;
	}
	Layout::__exit__(args);
	if (len == 1)
	{
		View &child = *py::cast<View*>(m_children[0]);
		ctrl().Initialize(child);
	}
	else if (len == 2)
	{
		View &child1 = *py::cast<View*>(m_children[0]);
		View &child2 = *py::cast<View*>(m_children[1]);
		if (m_horizontal)
			ctrl().SplitHorizontally(child1, child2, m_sashpos);
		else
			ctrl().SplitVertically(child1, child2, m_sashpos);
	}
}


void BookCtrlBase::doAdd(View & child)
{
	Item *item = (Item*)child.ptr()->GetClientData();
	if (Item::isInstance(item))
	{
		// 替换回原指针
		child.ptr()->SetClientData(&child);

		wxcstr caption = pyDictGet(item->m_kwargs, wxT("caption"), wxNoneString);
		ctrl().AddPage(child, caption);

		py::cast(item).dec_ref();
	}
	else
	{
		log_message(wxString::Format("Child of %s must be Item.", "Notebook"));
	}
}

View * BookCtrlBase::getPage(int n)
{
	if (n == -1)
	{
		n = getSelection();
	}
	return (View*)ctrl().GetPage(n)->GetClientData();
}


void init_containers(py::module & m)
{
	using namespace py::literals;

	auto className = "className"_a = None;
	auto style = "style"_a = None;
	auto styles = "styles"_a = None;
	auto wxstyle = "wxstyle"_a = 0;

	py::class_t<Vertical, Layout>(m, "Vertical")
		.def_init(py::init<pyobj, pyobj, pyobj>(), styles, className, style);

	py::class_t<Horizontal, Layout>(m, "Horizontal")
		.def_init(py::init<pyobj, pyobj, pyobj>(), styles, className, style);

	py::class_t<GridLayout, Layout>(m, "GridLayout")
		.def_init(py::init<int, int, int, int, pyobj, pyobj, pyobj>(),
			"rows"_a = 0, "cols"_a = 2, "vgap"_a = 0, "hgap"_a = 0,
			styles, className, style);

	py::class_t<FlexGridLayout, Layout>(m, "FlexGridLayout")
		.def_init(py::init<int, int, int, int, pyobj, pyobj, pyobj>(),
			"rows"_a = 0, "cols"_a = 2, "vgap"_a = 0, "hgap"_a = 0,
			styles, className, style)
		.def("AddGrowableRow", &FlexGridLayout::AddGrowableRow, "index"_a, "flex"_a = 0)
		.def("RemoveGrowableRow", &FlexGridLayout::RemoveGrowableRow, "index"_a)
		.def("AddGrowableCol", &FlexGridLayout::AddGrowableCol, "index"_a, "flex"_a = 0)
		.def("RemoveGrowableCol", &FlexGridLayout::RemoveGrowableCol, "index"_a)
		.def_property("flexDirection", &FlexGridLayout::GetFlexibleDirection, &FlexGridLayout::SetFlexibleDirection);

	py::class_t<ScrollView, Layout>(m, "ScrollView")
		.def_init(py::init<bool, pyobj, pyobj, pyobj>(),
			"horizontal"_a = false, styles, className, style);

	py::class_t<SplitterWindow, Layout>(m, "SplitterWindow")
		.def_init(py::init<bool, int, pyobj, pyobj, pyobj>(),
			"horizontal"_a = false, "sashpos"_a = 0, styles, className, style);

	py::class_t<StaticBox, Layout>(m, "StaticBox")
		.def_init(py::init<wxcstr, pyobj, pyobj, pyobj>(),
			"label"_a, styles, className, style)
		.def("getLabel", &StaticBox::getLabel)
		.def("setLabel", &StaticBox::setLabel)
		.def_property("label", &StaticBox::getLabel, &StaticBox::setLabel);

	py::class_t<BookCtrlBase, Layout>(m, "BookCtrlBase")
		.def("getPage", &BookCtrlBase::getPage, "n"_a = -1)
		.def("getPageCount", &BookCtrlBase::getPageCount)
		.def("setPageText", &BookCtrlBase::setPageText)
		.def("getPageText", &BookCtrlBase::getPageText)
		.def_property("index", &BookCtrlBase::getSelection, &BookCtrlBase::setSelection);

	py::class_t<Notebook, BookCtrlBase>(m, "Notebook")
		.def_init(py::init<int, pyobj, pyobj, pyobj>(), wxstyle, styles, className, style);

	py::class_t<Listbook, BookCtrlBase>(m, "Listbook")
		.def_init(py::init<int, pyobj, pyobj, pyobj>(), wxstyle, styles, className, style);
}

File: wxFEFactory/python/lib/ui/containers.py
from .view import View, Layout
from . import wx
from lib.exctypes import int32


class SizerLayout(Layout):
    def layout_child(self, child, style):
        flag = self.get_box_flag(style)
        weight = style.get('weight', 0)
        padding = style.get('padding', 5)
        self.sizer.Add(child.wxwindow, weight, flag, padding)

    def layout(self):
        self.sizer.Layout()

    def set_sizer(self, sizer):
        self.sizer = sizer
        self.SetSizer(sizer)

    def get_box_flag(self, style):
        """获取布局参数"""
        flag = 0
        if style.get('expand', False):
            flag |= wx.EXPAND

        padding_flag = style.get('padding-flag', 0)
        if padding_flag is not 0:
            if padding_flag is 1:
                flag |= wx.ALL
            else:
                if padding_flag & 0b1000:
                    flag |= wx.TOP
                if padding_flag & 0b0100:
                    flag |= wx.RIGHT
                if padding_flag & 0b0010:
                    flag |= wx.BOTTOM
                if padding_flag & 0b0001:
                    flag |= wx.LEFT

        vertical = style.get('vertical-align', None)
        if vertical is not None:
            if vertical == 'top':
                flag |= wx.ALIGN_TOP
            elif vertical == 'bottom':
                flag |= wx.ALIGN_BOTTOM
            elif vertical == 'middle':
                flag |= wx.ALIGN_CENTER_VERTICAL
            else:
                print("%s: %s not available" % ('vertical-align', vertical))

        align = style.get('align', None)
        if align is not None:
            if align == 'left':
                flag |= wx.ALIGN_LEFT
            elif align == 'right':
                flag |= wx.ALIGN_RIGHT
            elif align == 'center':
                flag |= wx.ALIGN_CENTER_HORIZONTAL
            else:
                print("%s: %s not available" % ('align', align))

        return flag


class SizerPanel(SizerLayout):
    wxtype = wx.Panel


class Vertical(SizerPanel):
    """垂直布局"""
    def onready(self):
        self.set_sizer(wx.BoxSizer(wx.VERTICAL))
        super().onready()


class Horizontal(SizerPanel):
    """水平布局"""
    def onready(self):
        self.set_sizer(wx.BoxSizer(wx.HORIZONTAL))
        super().onready()


class GridLayout(SizerPanel):
    """网格布局"""
    def __init__(self, rows=0, cols=2, vgap=0, hgap=0, **kwargs):
        self.rows = rows
        self.cols = cols
        self.vgap = vgap
        self.hgap = hgap
        super().__init__(**kwargs)

    def onready(self):
        self.set_sizer(wx.GridSizer(self.rows, self.cols, self.vgap, self.hgap))
        super().onready()


class FlexGridLayout(SizerPanel):
    """网格布局"""
    def __init__(self, rows=0, cols=2, vgap=0, hgap=0, **kwargs):
        self.rows = rows
        self.cols = cols
        self.vgap = vgap
        self.hgap = hgap
        super().__init__(**kwargs)

    def onready(self):
        sizer = wx.FlexGridSizer(self.rows, self.cols, self.vgap, self.hgap)
        sizer.SetNonFlexibleGrowMode(wx.FLEX_GROWMODE_SPECIFIED)
        self.set_sizer(sizer)
        super().onready()


class ScrollView(SizerLayout):
    wxtype = wx.ScrolledWindow

    def __init__(self, horizontal=False, wxstyle=int32(wx.HSCROLL | wx.VSCROLL).value, **kwargs):
        self.horizontal = horizontal
        super().__init__(wxstyle=wxstyle, **kwargs)

    def onready(self):
        self.SetScrollRate(5, 5)
        self.set_sizer(wx.BoxSizer(wx.HORIZONTAL if self.horizontal else wx.VERTICAL))
        super().onready()

    def layout(self):
        self.sizer.FitInside(self.wxwindow)
        self.Layout()


class SplitterWindow(Layout):
    """分割窗口"""
    wxtype = wx.SplitterWindow

    def __init__(self, horizontal=False, sashpos=0, **kwargs):
        self.horizontal = horizontal
        self.sashpos = sashpos
        Layout.__init__(self, **kwargs)

    def onready(self):
        super().onready()
        length = len(self.children)
        if length > 2:
            print('SplitterWindow不支持大于两个子元素')
            return
        if length is 1:
            self.Initialize(self.children[0])
        elif length is 2:
            child1 = self.children[0]
            child2 = self.children[1]
            if self.horizontal:
                self.SplitHorizontally(child1, child2, self.sashpos)
            else:
                self.SplitVertically(child1, child2, self.sashpos)


class StaticBox(SizerLayout):
    """静态框"""
    wxtype = wx.StaticBox

    def __init__(self, label, **kwargs):
        super().__init__(wxparams={'label': label}, **kwargs)

    def onready(self):
        sizer = wx.BoxSizer(wx.VERTICAL)
        sizer.InsertSpacer(0, 15)
        self.set_sizer(sizer)
        super().onready()


class BookCtrlBase(Layout):
    def layout_child(self, child, style):
        caption = child.extra['caption']
        self.AddPage(child.wxwindow, caption)

    def get_page(self, n=None):
        if n is None:
            n = self.GetSelection()
        else:
            count = self.GetPageCount()
            if n < 0:
                n += count
            if not 0 <= n < count:
                raise IndexError('list index out of range')

        return self.GetPage(n).GetHost()

    def set_onchange(self, onchange, reset=True):
        self.bind_event(self.wxevent, onchange, reset)

    @property
    def index(self):
        return self.GetSelection()

    @index.setter
    def index(self, n):
        self.SetSelection(n)


class Notebook(BookCtrlBase):
    wxtype = wx.Notebook
    wxevent = wx.EVT_NOTEBOOK_PAGE_CHANGED


class Listbook(BookCtrlBase):
    wxtype = wx.Listbook
    wxevent = wx.EVT_LISTBOOK_PAGE_CHANGED
